from django.contrib.auth.models import User, AnonymousUser
from rest_framework import serializers
from rest_framework.exceptions import ValidationError

from advertisements.models import Advertisement


class UserSerializer(serializers.ModelSerializer):
    """Serializer для пользователя."""

    class Meta:
        model = User
        fields = ('id', 'username', 'first_name',
                  'last_name',)


class AdvertisementSerializer(serializers.ModelSerializer):
    """Serializer для объявления."""

    creator = UserSerializer(
        read_only=True,
    )

    class Meta:
        model = Advertisement
        fields = ('id', 'title', 'description', 'creator',
                  'status', 'created_at', )

    def create(self, validated_data):
        """Метод для создания"""
        validated_data.pop('user')
        # Простановка значения поля создатель по-умолчанию.
        # Текущий пользователь является создателем объявления
        # изменить или переопределить его через API нельзя.
        # обратите внимание на `context` – он выставляется автоматически
        # через методы ViewSet.
        # само поле при этом объявляется как `read_only=True`
        validated_data["creator"] = self.context["request"].user
        return super().create(validated_data)

    def validate(self, data):
        """Метод для валидации. Вызывается при создании и обновлении."""
        ADV_STATUS = 'OPEN'
        ADV_LIMIT = 10
        if data.get('status', 'OPEN') == ADV_STATUS:
            amount_of_post = len(Advertisement.objects.filter(status=ADV_STATUS, creator=self.context['request'].user))
            if amount_of_post >= ADV_LIMIT:
                raise ValidationError(f"You can't have more than {ADV_LIMIT} opened posts.")
        return data
